pylzy/lzy/api/v1/provisioning.py

```diff
@@ -16,46 +76,178 @@ class GpuType(Enum):
     T4 = "T4"
 
 
-@dataclass(frozen=True)
+ScoreFunctionType = Callable[["Provisioning", "Provisioning"], float]
+
+
+def maximum_score_function(requested: "Provisioning", spec: "Provisioning") -> float:
+    """Score function that scores big machines the max.
+    Use it when you need the best spec available.
+    """
+
+    d_gpu = spec.gpu_count_final - requested.gpu_count_final
+    d_cpu = spec.cpu_count_final - requested.cpu_count_final
+    d_ram = spec.ram_size_gb_final - requested.ram_size_gb_final
+
+    return d_gpu * DEFAULT_WEIGHTS['gpu'] + d_cpu * DEFAULT_WEIGHTS['cpu'] + d_ram * DEFAULT_WEIGHTS['ram']
+
+
+def minimum_score_function(requested: "Provisioning", spec: "Provisioning") -> float:
+    """Score function that scores small machines the max.
+    Use it when you need smallest spec available that matches you requirements.
+
+    """
+
+    max_score = maximum_score_function(requested=requested, spec=spec)
+    return 1.0 / max_score if max_score else float("inf")
+
+
+@dataclasses.dataclass(frozen=True)
 class Provisioning:
-    cpu_type: Optional[str] = None
-    cpu_count: Optional[int] = None
-
-    gpu_type: Optional[str] = None
-    gpu_count: Optional[int] = None
-
-    ram_size_gb: Optional[int] = None
-
-    @staticmethod
-    def default() -> "Provisioning":
-        return Provisioning(
-            cpu_type=CpuType.CASCADE_LAKE.value,
-            cpu_count=2,
-            gpu_type=GpuType.NO_GPU.value,
-            gpu_count=0,
-            ram_size_gb=2,
-        )
+    cpu_type: StringRequirement = None
```

Comment:
Why default is None here.
Mb Any?

Reply:
I had this discussion earlier with @mrMakaronka at this PR.
tldr None is a synonym for "user does not pass anything" and `Any` - is an explicit value and we need to maintain this difference for a transparent merging of values.

Reply:
Ok